dkc/core/models/folder.py

```diff
@@ -59,8 +58,56 @@ class Meta:
         'self', on_delete=models.DO_NOTHING, blank=True, null=True, related_name='+'
     )
 
-    # # Prevent deletion of quotas while a folder references them
-    # quota = models.ForeignKey(Quota, on_delete=models.PROTECT)
+    def resolve_quota(self) -> Tuple[int, int]:
+        """
+        Resolve the quota for a given folder.
+
+        Returns the value as a tuple of the form (bytes_used, bytes_allowed).
+        """
+        folder_quota = self.root_folder.quota
+        if folder_quota.allowed is None:
+            # This indicates that this folder's quota is its owning user's quota
+            owner_quota = self.root_folder.owner.quota
+            return owner_quota.used, owner_quota.allowed
+        return folder_quota.used, folder_quota.allowed
+
+    @transaction.atomic
+    def increment_quota(self, amount: int) -> None:
+        """
+        Increments the quota(s) associated with this folder.
+
+        This function increments (or decrements, if ``amount`` is negative) the usage
+        values associated with this folder. The used amount tracked on the folder itself is
+        incremented. Additionally, if this folder has a user-assigned quota, that quota
+        usage value is also incremented. If either of these increments would exceed the max
+        allotment for this folder, a ``ValidationError`` is raised and the transaction
+        is rolled back.
+        """
+        if amount == 0:
+            return
+
+        folder_quota = self.root_folder.quota
+        folder_quota.used = models.F('used') + amount
+        folder_quota.save()
+
+        if folder_quota.allowed is None:
+            user_quota = self.root_folder.owner.quota
+            user_quota.used = models.F('used') + amount
+            user_quota.save()
+            if amount > 0:
+                user_quota.refresh_from_db()
+                if user_quota.used > user_quota.allowed:
+                    raise ValidationError(
+                        'User size quota would be exceeded: '
+                        f'{user_quota.used}B > {user_quota.allowed}B.'
+                    )
+        elif amount > 0:
+            folder_quota.refresh_from_db()
```

Comment:
Please advise on best practices here -- because I'm using an `F` expression, `refresh_from_db` needs to be called before the updated attributes will have their normally-typed values. I chose to only call `refresh_from_db` internally here if it's strictly necessary, for the sake of performance, but callers of this function would then have to be concerned with refreshing the objects themselves if they need to use these properties again (which they probably won't).